### ISSUE TITLE
Encoding error while creating a package since conan 1.2.0+ with Python 2.7.14

### ISSUE THREAD

**Issue by Aalmann** [conan-io/conan#2802]
Not sure how to reproduce, but I got a stacktrace when trying to package. Conan 1.0.4 and 1.1.1 works without errors. 
But with conan 1.2.0 and 1.2.3 I got the following output. (Conan installed with pip in Python 2.7.14)
```
Traceback (most recent call last):
  File "c:\python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\command.py", line 1187, in run
    method(args[0][1:])
  File "c:\python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\command.py", line 226, in create
    test_build_folder=args.test_build_folder)
  File "c:\python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\conan_api.py", line 61, in wrapper
    return f(*args, **kwargs)
  File "c:\python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\conan_api.py", line 323, in create
    keep_build=keep_build)
  File "c:\python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\manager.py", line 395, in install
    installer.install(deps_graph, profile.build_requires, keep_build)
  File "c:\python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\installer.py", line 264, in install
    root_conanfile)
  File "c:\python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\installer.py", line 330, in _build
    keep_build, profile_build_requires, flat, deps_graph)
  File "c:\python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\installer.py", line 385, in _build_package
    builder.package()
  File "c:\python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\installer.py", line 134, in package
    install_folder, self._out)
  File "c:\python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\packager.py", line 79, in create_package
    _create_aux_files(install_folder, package_folder, conanfile, copy_info)
  File "c:\python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\packager.py", line 100, in _create_aux_files
    save(os.path.join(package_folder, CONAN_MANIFEST), str(digest))
UnicodeEncodeError: 'ascii' codec can't encode character u'\xa0' in position 203792: ordinal not in range(128)

ERROR: 'ascii' codec can't encode character u'\xa0' in position 203792: ordinal not in range(128)

```
Currently I have no idea how to reproduce it with a small and simple package. But probably you have an idea or a fix by checking the stacktrace.

Best Aalmann

To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by Aalmann** [comment]
Additional information concerning the package directory.

- 574 MB
- 12,072 files, 651 directories

**Comment by Aalmann** [comment]
@memsharded 
I can reproduce it. If the filename contains special characters like German `ä,ö,ü` or like in this case a `nbsp`, then the `conans.util.files.md5sum` method (which is called in `FileTreeManifest.create`) throws an exception.
Use the following script to reproduce it (not sure if the `nbsp` is pasted correctly, but it should also fail with the spcial characters):
```
#!/bin/bash -e

echo removing old packages
conan remove Module_Z/1.0.0@*/* -f || echo nothing to remove

echo
echo
echo Creating a new Module_Z recipe
echo ------------------------------
rm -rf Module_Z
mkdir Module_Z
cd Module_Z
conan new Module_Z/1.0.0 --sources
touch "src/nbsp_between_here and_here.h"
touch "src/üäö.h"
cd ..

echo
echo
echo Creating package of Module_Z
echo ----------------------------
conan create ./Module_Z user/testing

```

Btw. I have no idea why there is a `nbsp` within the filename. But I can't change it, because it is an "external" program which is packaged. But I think special characters in filenames will always be possible. So conan has to handle it. And again, it works until conan 1.1.1 but not with 1.2+

**Comment by memsharded** [comment]
Hi @Aalmann 

Thanks for your detailed report. The truth is that unicode filenames are a nightmare for Python2 in general, and of course it affects conan. For non-ascii filenames, it is recommended Python3 based installations.

However, I am checking if it could be a regression, if it was working in previous versions could be.

**Comment by memsharded** [comment]
Able to reproduce regression, investigating.

**Comment by memsharded** [comment]
This is one of the weirdest bugs I have investigated. Had to bisect the history and be checking.

- It happens that pylint **changed** the default encoding, which is recognized to be a bad practice: https://anonbadger.wordpress.com/2015/06/16/why-sys-setdefaultencoding-will-break-code/
- We have recently "removed" the pylint direct dependency, and now runs as a subprocess. Consequently, it is not able to change the encoding any longer.
- That encoding change was saving some encoding errors. 
- PR #2806 seems to solve it. It would be great if you could do that little change to your codebase, then run from sources against your project to validate.

Thanks again.

**Comment by Aalmann** [comment]
Sadly it doesn't work. I think it is because `md5sum` is called instead of `md5`. `md5sum` uses the `_generic_algorithm_sum`.

**Comment by memsharded** [comment]
Not fully sure about that, in your traces (https://github.com/conan-io/conan/issues/2802#issuecomment-384013129), the ``md5`` is the one being shown, not ``md5sum``, could you please clarify this? Thanks!

**Comment by Aalmann** [comment]
This was "just" the ouput of pydev which was unable to resolve the value.
Because pydev was unable to show the content of digest, I have locally modified the loop in `FileTreeManifest.create` 
```
        with open("D:\\.conan\\mvqlow\\files.txt", 'w') as f:
            f.write("%s\n" % files)
            file_dict = {}
            for name, filepath in files.items():
                file_dict[name] = md5sum(filepath)
                f.write("%s : %s \n" % (name, file_dict[name]))
```
by adding a print to file (of `files` and `file_dict[name]` and `name`) to be able to get the 'bad' filename. The loop itself stops when hitting the 'bad' file and no further debug break point will be hit after that loop in this method.
So I was able to compare `files` with the incomplete `md5sum` result.
Sorry for confusing.

**Comment by memsharded** [comment]
No problem :) 

Still, sorry, so what is the correct trace and error? Do you still seeing it fail with the "utf-8"?

Did you already connected to CppLang slack team (#conan channel), or other way you want to chat? Feel free to write to info@conan.io to suggest, it could be useful to discuss this kind of issues.

**Comment by Aalmann** [comment]
Ignore comment 2.

`_create_aux_files` hit the breakpoint at line 100 `save` which is completely confusing, because the error itself occurs a line before. But there is no exception thrown. I added a `try ... except` around `FileTreeManifest.create`. So the stacktrace and error is still the same as described in comment 1 (with and without the `'UTF-8'`.
When changing to `save(os.path.join(package_folder, CONAN_MANIFEST), unicode(digest))` will produce the following stacktrace
```
Traceback (most recent call last):
  File "c:\Python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\command.py", line 1187, in run
    method(args[0][1:])
  File "c:\Python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\command.py", line 226, in create
    test_build_folder=args.test_build_folder)
  File "c:\Python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\conan_api.py", line 61, in wrapper
    return f(*args, **kwargs)
  File "c:\Python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\conan_api.py", line 323, in create
    keep_build=keep_build)
  File "c:\Python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\manager.py", line 395, in install
    installer.install(deps_graph, profile.build_requires, keep_build)
  File "c:\Python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\installer.py", line 264, in install
    root_conanfile)
  File "c:\Python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\installer.py", line 330, in _build
    keep_build, profile_build_requires, flat, deps_graph)
  File "c:\Python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\installer.py", line 385, in _build_package
    builder.package()
  File "c:\Python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\installer.py", line 134, in package
    install_folder, self._out)
  File "c:\Python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\packager.py", line 79, in create_package
    _create_aux_files(install_folder, package_folder, conanfile, copy_info)
  File "c:\Python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\client\packager.py", line 104, in _create_aux_files
    save(os.path.join(package_folder, CONAN_MANIFEST), unicode(digest))
  File "c:\Python\2.7.14\lib\site-packages\conan-1.2.3-py2.7.egg\conans\util\files.py", line 113, in save
    handle.write(to_file_bytes(content))
UnicodeEncodeError: 'ascii' codec can't encode character u'\xa0' in position 203793: ordinal not in range(128)
```

**Comment by memsharded** [comment]
This is what is really confusing to me. It is not failing that way on my machine. Different behavior in different computers... encodings are terrible.

Maybe it would be worth to print the default encoding, just in case, just before the error, something like:

```
   print sys.getdefaultencoding()
   save(os.path.join(package_folder, CONAN_MANIFEST), str(digest))
```

**Comment by Aalmann** [comment]
The output is `ascii`
I think the problem is probably around `gather_files` in `conans.model.manifest`
I have modified line 31 to
```
            if os.path.exists(abs_path):
                file_dict[rel_path] = abs_path.encode()
```
which throws the same UnicodeEncodeError.

**Comment by memsharded** [comment]
What about?
```
file_dict[rel_path] = abs_path.encode("utf-8")
```

I keep working on it too, by now most of my day, and I can't figure out something that makes sense :(

This is really the kind of thing that has pushed some users to upgrade to Python3. It seems it is too problematic issue, like playing whack-a-mole, too many things that are out of control (system language, terminal encoding, python installation encoding...)

**Comment by Aalmann** [comment]
Hi @memsharded 

tested your PR #2807 locally with my "bad" package.
Result: It works with my "bad" package and with the `äöü` example above.

Many thanks for fixing.

**Comment by memsharded** [comment]
This should be fixed in 1.3.3, please check and report otherwise. Thanks!

**Comment by ericriff** [comment]
Strangely enough, I came across a very similar error today. I can build my package locally on Ubuntu16+conan 1.21 and also Ubuntu 18 + conan 1.29.0. On our CI servers, we rock the same configs but on dockers.
On the docker images, I get
```
DEBUG :packager.py    [112]: PACKAGE: Creating config files to /home/ubuntu/.conan/data/pkg/1.0/usr/stable/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 [2020-09-30 23:12:38,585]
Traceback (most recent call last):
  File "conan/conans/client/command.py", line 1947, in run
  File "conan/conans/client/command.py", line 358, in create
  File "conan/conans/client/conan_api.py", line 81, in wrapper
  File "conan/conans/client/conan_api.py", line 364, in create
  File "conan/conans/client/cmd/create.py", line 43, in create
  File "conan/conans/client/cmd/test.py", line 38, in install_build_and_test
  File "conan/conans/client/manager.py", line 68, in deps_install
  File "conan/conans/client/installer.py", line 308, in install
  File "conan/conans/client/installer.py", line 332, in _build
  File "conan/conans/client/installer.py", line 398, in _handle_node_cache
  File "conan/conans/client/installer.py", line 442, in _build_package
  File "conan/conans/client/installer.py", line 212, in build_package
  File "conan/conans/client/installer.py", line 161, in _package
  File "conan/conans/client/packager.py", line 91, in run_package_method
  File "conan/conans/client/packager.py", line 125, in _create_aux_files
  File "conan/conans/model/manifest.py", line 110, in save
  File "conan/conans/util/files.py", line 183, in save
  File "conan/conans/util/files.py", line 201, in to_file_bytes
UnicodeEncodeError: 'utf-8' codec can't encode characters in position 1257517-1257518: surrogates not allowed

ERROR: 'utf-8' codec can't encode characters in position 1257517-1257518: surrogates not allowed
```

And I cant reproduce it locally or figure out what is going on.
We are using python 3.5 on U16 and 3.6 on U18.

Any suggestions to debug this?
